Dash DIP2 transactions support

gladcow opened the pull request:
This PR adds support for new Dash transactions format as presented in DIP2: (https://github.com/dashpay/dips/blob/master/dip-0002.md) and allows Trezor to fully support all required features, including managing of Deterministic Masternodes by Masternode Owners, which should significantly streamline the process for such Trezor users.

Unlike the previous attempt to add this functionality (https://github.com/trezor/trezor-common/pull/283, https://github.com/trezor/trezor-core/issues/484), this PR implements it:
 * without any changes to the protocol (no changes in trezor-common);
 * with full verification of the new transaction data by wallet code and user confirmation for all used parameters;
 * with all Dash-specific code moved to a separate module;

Commit:
Dash version 3 transactions tests

Commit:
Special processing of Dash version 3 transactions

Commit:
Add transaction extra data support in btc.sign_tx

Commit:
Ability to preload an external txes which are not inputs in btc.sign_tx

malinah committed:
core: redesign multiprofile support, add profile verbosity

andrewkozlik committed:
core/cbor: move CBOR from cardano to common

- add support for text strings and boolean values
- add support negative integers and decoding maps
- fix decoding of short strings
- encode maps canonically
- add unit tests for decoding
- sort maps lexicographically by encoded key

Commit:
Check Dash special tx output ownership

romanz committed:
core/secp256k1_zkp: refactor API into a Context class

In order to prevent frequent memory allocations, the user can create a single
Context object and re-use it between subsequent cryptographic operations.

Scotty0448 committed:
common/defs: add Ritocoin blockbook host (#198)

thephez commented on core/src/apps/wallet/sign_tx/dash.py:
```diff
@@ -0,0 +1,470 @@
+    version = tx.version
+    dip2_type = version >> 16
+    version &= 0xffff
+    return (version is 3) and (dip2_type > 0)
```
Maybe missing something, but shouldn't `(version is 3)` be `(version >= 3)` since future updates may increase the version number?

gladcow replied:
We don't know anything about version 4 transactions and we don't know really if it is a correct way to process this version 4 transactions,  so we call this functionality only for version 3 transactions. It is the same condition like in the Core,  here,  for example: https://github.com/dashpay/dash/blob/master/src/evo/specialtx.cpp#L22

UdjinM6 commented on core/src/apps/wallet/sign_tx/signing.py:
```diff
@@ -164,6 +170,12 @@ class SigningError(ValueError):
+        if dash.is_dip2_tx(tx) and txo.address_n is None:
+            raise SigningError(
+                FailureType.ActionCancelled,
+                "Cannot send Dash special transaction to external address",
```
Would be nice to have some tests for this.

ph4r05 committed:
xmr: get_address - pagination, integrated address

- pagination added to show_address
- integrated address supported
- QR code fixed (prefix with monero: as GUI does)

seanmcgary committed:
common/defs: add ASK (#200)

tsusanka committed:
submodules: temporary set ElementsProject/secp256k1-zkp submodule a fork due to missing reference

See https://github.com/ElementsProject/secp256k1-zkp/issues/68

tsusanka committed:
tests: remove progress bar from curl

It makes CI logs nicer.

tsusanka committed:
Revert "submodules: temporary set ElementsProject/secp256k1-zkp submodule a fork due to missing reference"

This reverts commit 088760fa1e42bca95e9362706b512505153d2356.

tsusanka committed:
xmr: get_address - pagination, integrated address (#201)

xmr: get_address - pagination, integrated address

matejcik committed:
git subrepo push common

subrepo:
  subdir:   "common"
  merged:   "31a0073c"
upstream:
  origin:   "git+ssh://git@github.com/trezor/trezor-common"
  branch:   "master"
  commit:   "31a0073c"
git-subrepo:
  version:  "0.4.0"
  origin:   "https://github.com/ingydotnet/git-subrepo.git"
  commit:   "5d6aba9"

Scotty0448 committed:
common/defs: update sign message headers for RITO and RVN

prusnak committed:
common/defs: add to support.json, regenerate

prusnak committed:
core/cbor: fix string conversion to work with bytearrays

romanz committed:
common/defs: add Elements regtest support (#213)

Commit:
Test for exernal output in Dash DIP2 tx

UdjinM6 commented on python/trezorlib/tests/device_tests/test_msg_signtx_dash.py:
```diff
@@ -937,3 +939,39 @@ def test_send_dash_dip2_prouprevtx(self):
+            script_type=proto.InputScriptType.SPENDADDRESS,
+        )
+        out1 = proto.TxOutputType(
+            address="XpTc36DPAeWmaueNBA9JqCg2GC8XDLKSYe",
```
That's not a testnet addresss

UdjinM6 commented on python/trezorlib/tests/device_tests/test_msg_signtx_dash.py:
```diff
@@ -937,3 +939,39 @@ def test_send_dash_dip2_prouprevtx(self):
+            )
+
+        assert exc.value.args[0] == proto.FailureType.DataError
+        assert exc.value.args[1].endswith("Invalid address type")
```
Should be `"Cannot send Dash special transaction to external address"`

gladcow replied:
Yes,  you are right,  I've tested other case really. I've fixed it,  testing it now,  will commit in several minutes

gladcow commented:
I'm not checking exception params now because they are overwritten here: https://github.com/gladcow/trezor-firmware/blob/dash_special/python/trezorlib/client.py#L221

UdjinM6 reviewed (approved):
LGTM 👍

gladcow commented:
I've squashed the fixes commits to pull it in Trezor repo.

Commit:
style fix

Commit:
move address_pkh_from_keyid function to dash.py
